src/twisted/internet/_sslverify.py

```diff
@@ -1684,9 +1684,9 @@ def _verifyCallback(conn, cert, errno, depth, preverify_ok):
 
         if self.enableSessions:
             name = "%s-%d" % (reflect.qual(self.__class__), _sessionCounter())
-            sessionName = md5(networkString(name)).hexdigest()
+            sessionName = sha256(networkString(name)).digest()
 
-            ctx.set_session_id(sessionName.encode('ascii'))
+            ctx.set_session_id(sessionName)
```

Comment:
Is there a reason not to just generate random session identifiers: `os.urandom(16)`?

Reply:
From what I can tell, not really. The OpenSSL docs mention that there are problems if this is not unique to the application (https://www.openssl.org/docs/man1.0.2/ssl/SSL_CTX_set_session_id_context.html), but I don't see any reason why a random byte sequence and this hash would be more/less unique than each other.

Presumably the current strings (a counter based on how many sessions have been created), don't need to be reproducible -- the only benefit that I can think of to using the hash here is that that the first _n_ session IDs used by Twisted will always be the same (but that could also be spun as a downside). The docs also mention that conflicting IDs can cause problems, so I'm curious why this hasn't been a problem before w/ multiple twisted clients in different processes negotiating session IDs w/ the same server.

I'll update this to use random numbers, which I prefer as a solution, but I'm still uneasy due to not fully understanding why this was a hash in the first place. The original commit appears to be lost to time (oldest I can find w/ this file is a merge of an svn branch which I can't find).

Reply:
Here's the first committed revision that included this code:

https://github.com/twisted/vertex/commit/389656bed398f2d4940cb3de1e2c10ee6204c7aa#diff-e2026d749e1abadafbd27bb502844bdcR670

It doesn't shed much light, though.

Reply:
I spent some time digging into OpenSSL docs and source code (*not* my favorite pastime, mind) to try to figure out what's actually going on here. It turns out that what we're dealing with here is not the actual TLS SessionID, but the OpenSSL "Session ID Context". (That is, the PyOpenSSL `set_session_id` method is a misnomer, as it in turn calls `SSL_CTX_set_session_id_context`). The OpenSSL manpages do a poor job of explaining what this is; the most coherent explanation I could find is in the O'Reilly OpenSSL book:

> All session must have a session ID context. For the server, session caching is disabled by default unless a call to `SSL_CTX_set_session_id_context` is made. The purpose of the session ID context is to make sure the session is being reused for the same purpose for which it was created. For instance, a session created for an SSL web server should not be automatically allowed for an SSL FTP server. Along the same lines, we can use session ID contexts to exercise a finer-grained control over sessions within our application. For example, authenticated clients could have a different session ID context than unauthenticated ones. The context itself can be any data we choose...

So, to put it another way, we can think of the "session ID context" as a namespace for OpenSSL sessions. As the quote suggests, we probably don't normally want to allow sessions to be resumed across different `SSL_CTX` instances, so making this a unique value per context, by default, seems preferable. As Marcus said, using random generation (assuming a well-behaved CSPRNG) with a length of 32 bytes will accomplish this with very high probability.